File: 백준/Gold/15789. CTP 왕국은 한솔 왕국을 이길 수 있을까？/CTP 왕국은 한솔 왕국을 이길 수 있을까？.py
def find(x):
    if root[x] == x:
        return x
    else:
        root[x] = find(root[x])
        return root[x]

def unionNode(x, y):
    rootX = find(x)
    rootY = find(y)
    if rootX != rootY:
        root[rootX] = rootY

def main():
    # import sys
    # input = sys.stdin.read
    # # data = input().split()
    
    # # idx = 0
    
    # # n = int(data[idx])
    # # idx += 1
    # # m = int(data[idx])
    # # idx += 1
    
    n, m = map(int, input().split())
    global root
    root = [i for i in range(n+1)]
    
    for _ in range(m):
        # a = int(data[idx])
        # idx += 1
        # b = int(data[idx])
        # idx += 1
        a, b = map(int, input().split())
        unionNode(a, b)
    ctp, hansol, k = map(int, input().split())
    # ctp = int(data[idx])
    # idx += 1
    # hansol = int(data[idx])
    # idx += 1
    # k = int(data[idx])
    # idx += 1
    
    ctp_root = find(ctp)
    hansol_root = find(hansol)
    
    arr = [0] * (n+1)
    
    for i in range(1, n+1):
        idx = find(i)
        arr[idx] += 1
    
    ctp_power = arr[ctp_root]
    arr[ctp_root] = 0
    arr[hansol_root] = 0
    
    arr.sort(reverse=True)
    
    for i in range(k):
        ctp_power += arr[i]
    
    print(ctp_power)

if __name__ == "__main__":
    main()

File: 백준/Gold/15789. CTP 왕국은 한솔 왕국을 이길 수 있을까？/CTP 왕국은 한솔 왕국을 이길 수 있을까？.cc
#include <bits/stdc++.h>
#define pii pair<int, int>
#define ll long long
//CTP 왕국은 한솔 왕국을 이길 수 있을까? 
using namespace std;
vector<int> root;

int find(int a){
    if(root[a]==a) {
        return a;
    }
    else {
        return root[a]=find(root[a]);
    }
}
void unionNodes(int a, int b){
    int rootA = find(a);
    int rootB = find(b);

    if(rootA!=rootB){
        root[rootA] = rootB;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int n, m; // 왕국의 수 , 동맹 관계 수 
    // 최대 k개의 동맹, gtp 왕국은 한솔 왕국과 동맹인 왕국과는 동맹을 맺을 수 없음 (아무리 최장이여도 한솔왕국이 포함되어있다면 안됨), 
    // 최대 k번 동맹을 맺을 수 있지만, k번을 다 안해도 됨
    cin >> n >> m;
    
    root.resize(n+1,0);

    // 초기화 
    for(int i=1; i<=n; i++){
        root[i]=i; 
    }

    int a,b;
    // 동맹 입력 받기 
    for(int i=0; i<m; i++){
        cin >> a >> b;
        unionNodes(a,b);
    }
    vector<pii> temp;
    int CTP, hansol, k ;
    cin >> CTP >> hansol >> k;

   
    int ctp_root = find(CTP);
    int hansol_root = find(hansol);

    vector<int> arr(n+1,0);

    for(int i=1; i<=n; i++){
        int idx = find(i);
        arr[idx]++;
    }
    

    
    // ctp와 연결된 집합들의 크기 
    int ctp_power = arr[ctp_root];
  //  cout << ctp_power << "\n";
    arr[ctp_root]=0; arr[hansol_root]=0;


    sort(arr.begin(), arr.end(), greater<int>());
    // for(int i=0; i<=n; i++){
    //    // cout << arr[i] << " ";
    // }

    for(int i=0; i< k; i++){
        ctp_power += arr[i];
    }
    cout << ctp_power<<"\n";

    return 0;
}